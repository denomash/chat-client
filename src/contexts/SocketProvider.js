import React, { useContext, useEffect, useState } from "react";
import io from "socket.io-client";

const SocketContext = React.createContext();

const useSocket = () => useContext(SocketContext);

const ENDPOINT = "localhost:5000";

const SocketProvider = ({ username, children }) => {
  const [socket, setSocket] = useState();

  useEffect(() => {
    const newSocket = io(ENDPOINT, {
      transports: ["websocket"],
      query: { id: username },
    });
    setSocket(newSocket);

    return () => newSocket.close();
  }, [username]);

  return (
    <SocketContext.Provider value={socket}>{children}</SocketContext.Provider>
  );
};

export { useSocket, SocketProvider };
