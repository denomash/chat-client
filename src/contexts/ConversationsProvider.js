import React, { useCallback, useContext, useEffect, useState } from "react";
import usePersistData from "../hooks/usePersistData";
import { useSocket } from "./SocketProvider";

const ConversationsContext = React.createContext();

const useConversations = () => {
  return useContext(ConversationsContext);
};

const ConversationsProvider = ({ username, children }) => {
  const [conversations, setConversations] = usePersistData("conversations", []);
  const [selectedConversationIndex, setSelectedConversationIndex] = useState(0);

  const socket = useSocket();

  const createConversations = (recipients) => {
    setConversations((prevConversations) => {
      return [...prevConversations, { recipients, messages: [] }];
    });
  };

  const addMessageToConversation = useCallback(
    ({ recipient, recipients, text, sender }) => {
      setConversations((prevConversations) => {
        let convoExists = false;
        let newMessage = { text, sender };
        const newConversations = prevConversations.map((conversation) => {
          const recievers = conversation.recipients.map(
            (recipient) => recipient.username
          );

          if (recievers.includes(recipient)) {
            convoExists = true;
            return {
              ...conversation,
              messages: [...conversation.messages, newMessage],
            };
          }
          return conversation;
        });

        // Check if conversation exists
        // If exists update
        // Otherwise update existing conversations with the new one
        if (convoExists) {
          return newConversations;
        } else {
          // This mostly happens on the new tab we are starting a new conversation with
          return [
            ...prevConversations,
            { recipients, recipient, messages: [newMessage] },
          ];
        }
      });
    },
    [setConversations]
  );

  useEffect(() => {
    if (socket == null) return;

    socket.on("receive-message", addMessageToConversation);
    return () => socket.off("receive-message");
  }, [socket, addMessageToConversation, conversations]);

  const sendMessage = (recipient, recipients, text) => {
    socket.emit("send-message", { recipient, recipients, text });
    addMessageToConversation({ recipients, recipient, text, sender: username });
  };

  const convos = conversations.map((convo, i) => {
    const messages = convo.messages.map((message) => {
      const fromMe = username === message.sender;
      return { ...message, fromMe };
    });

    return {
      ...convo,
      messages,
    };
  });

  return (
    <ConversationsContext.Provider
      value={{
        conversations: convos,
        createConversations,
        selectedConversation: convos[selectedConversationIndex],
        selectConversationIndex: setSelectedConversationIndex,
        sendMessage,
      }}
    >
      {children}
    </ConversationsContext.Provider>
  );
};

export { useConversations, ConversationsProvider };
