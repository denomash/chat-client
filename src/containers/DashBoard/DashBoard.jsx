import { Box } from "@mui/material";
import React from "react";
import { useLocation } from "react-router-dom";
import queryString from "query-string";

// Hooks
import { SideBar } from "../../components/SideBar/SideBar";
import { SocketProvider } from "../../contexts/SocketProvider";

// Components
import {
  ConversationsProvider,
  useConversations,
} from "../../contexts/ConversationsProvider";
import Chat from "../Chat/Chat";

const Container = ({ id }) => {
  const { selectedConversation } = useConversations();

  return (
    <>
      <SideBar username={id} />
      {selectedConversation && <Chat username={id} />}
    </>
  );
};

export const DashBoard = () => {
  const history = useLocation();

  const { id } = queryString.parse(history.search);

  return (
    <Box height="100vh" display="flex">
      <SocketProvider username={id}>
        <ConversationsProvider username={id}>
          <Container id={id} />
        </ConversationsProvider>
      </SocketProvider>
    </Box>
  );
};
