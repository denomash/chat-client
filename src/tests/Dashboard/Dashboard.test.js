import { Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { createMemoryHistory } from "history";

import { DashBoard } from "../../containers/DashBoard/DashBoard";
import { ConversationsProvider } from "../../contexts/ConversationsProvider";
import { SocketProvider } from "../../contexts/SocketProvider";

const Container = () => {
  const history = createMemoryHistory();

  return (
    <Router history={history}>
      <SocketProvider username="test">
        <ConversationsProvider username="test">
          <DashBoard username="test" />
        </ConversationsProvider>
      </SocketProvider>
    </Router>
  );
};

it("renders correctly when there are no items", () => {
  const tree = renderer.create(<Container />).toJSON();
  expect(tree).toMatchSnapshot();
});
