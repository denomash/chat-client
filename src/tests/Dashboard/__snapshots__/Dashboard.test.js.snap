// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div
  className="sc-gsDKAQ hkzyiR MuiBox-root"
>
  <div
    className="sc-gsDKAQ makeStyles-root-1 MuiBox-root"
    style={
      Object {
        "borderColor": "#1976d2",
        "borderRight": "solid 2px",
      }
    }
  >
    <div
      className="sc-gsDKAQ fpEaOL MuiBox-root"
    >
      <div
        className="sc-gsDKAQ lfLPC MuiBox-root"
      >
        <div
          className="MuiTabs-root-gEPFpJ hpisFx MuiTabs-root"
        >
          <div
            className="MuiTabs-scroller-ezHawR gCPLWN MuiTabs-scroller MuiTabs-fixed"
            onScroll={[Function]}
            style={
              Object {
                "marginBottom": -0,
                "overflow": "hidden",
              }
            }
          >
            <div
              aria-label="Chat"
              aria-orientation={null}
              className="MuiTabs-flexContainer-jDclAr ceXBBS MuiTabs-flexContainer"
              onKeyDown={[Function]}
              role="tablist"
            >
              <button
                aria-controls={null}
                aria-selected={true}
                className="MuiButtonBase-root-KjFju gUleGd MuiButtonBase-root MuiTab-root-bdqkgu isUQNB MuiTab-root MuiTab-textColorPrimary Mui-selected"
                disabled={false}
                id={null}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="tab"
                style={
                  Object {
                    "width": 250,
                  }
                }
                tabIndex={0}
                type="button"
              >
                Conversations
                <span
                  className="MuiTabs-indicator-csbZOi eudFOA MuiTabs-indicator"
                  style={Object {}}
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby={null}
        className="MuiTabPanel-root-drROWA jcHFQn MuiTabPanel-root"
        hidden={false}
        id={null}
        role="tabpanel"
        style={
          Object {
            "padding": 0,
          }
        }
      >
        <div
          className="sc-gsDKAQ iYUFjd MuiBox-root"
        >
          <div
            className="MuiTabs-root-gEPFpJ cdsNmx MuiTabs-root MuiTabs-vertical"
          >
            <div
              className="MuiTabs-scroller-ezHawR gCPLWN MuiTabs-scroller MuiTabs-fixed"
              onScroll={[Function]}
              style={
                Object {
                  "marginRight": -0,
                  "overflow": "hidden",
                }
              }
            >
              <div
                aria-label="Conversations"
                aria-orientation="vertical"
                className="MuiTabs-flexContainer-jDclAr jPZrdK MuiTabs-flexContainer MuiTabs-flexContainerVertical"
                onKeyDown={[Function]}
                role="tablist"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="sc-gsDKAQ iJVHIA MuiBox-root"
    >
      <div
        className="sc-gsDKAQ dZUCUE MuiBox-root"
      >
        <p
          className="MuiTypography-root-bWKFam fLRtBW MuiTypography-root MuiTypography-body1 makeStyles-userInfo-2"
        >
          Your username:
        </p>
        <p
          className="MuiTypography-root-bWKFam XRxud MuiTypography-root MuiTypography-body1"
        />
      </div>
      <button
        className="MuiButtonBase-root-KjFju gUleGd MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root MuiButton-root-gygCtr fSFcLn"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "borderRadius": 0,
            "marginTop": 10,
            "width": "100%",
          }
        }
        tabIndex={0}
        type="submit"
      >
        New Conversation
      </button>
    </div>
  </div>
</div>
`;
